### ISSUE TITLE
DOC: "Scalar arrays" (or 0-dimensional arrays)?

### ISSUE THREAD

**Issue by EricCousineau-TRI** [numpy/numpy#17744]
## Documentation

I skimmed through the docs here:
https://numpy.org/devdocs/user/index.html

As well as for the version that I've been using on Ubuntu 18.04:
https://numpy.org/doc/1.13/user/quickstart.html

The docs talk about array scalars here:
https://numpy.org/devdocs/reference/arrays.scalars.html

So this all is great when dealing with different `dtype`s, etc., esp. for `dtype=object` for our use.

However, things can get confusing when you see an array of size 1 but dimension 0, e.g.:
```
>>> import numpy as np
>>> np.__version__
'1.19.2'
>>> x = np.array(1.0)
>>> x.shape
()
>>> x.item()
1.0  # float, not np.float64
```

Is this mentioned anywhere in the docs with a more succinct name?

**Comment by seberg** [comment]
@EricCousineau-TRI do you have a question about the actual behaviour/definitions? I agree that there probably should be documentation explaining this more somewhere, not sure where immediately.

0-dimensional array is synonymous with scalar array. 1-Sized arrays unfortunately sometimes are also special handled (for things like `float(arr)` and `int(arr)`), but are not quite the same as 0-dimensional (scalar) arrays.

Array-scalars are the scalar versions, which unlike arrays are immutable. But it is important to note that NumPy is very agressive at turning scalar arrays into array scalars. (i.e. turn 0-D arrays into scalars.) That behaviour can be bothersome if you are working with `object` arrays.

**Comment by anthonyhvo12** [comment]
@seberg @EricCousineau-TRI It appears that 0-D array is briefly explained in https://numpy.org/devdocs/reference/arrays.ndarray.html#calculation. However, I think that should be mentioned earlier in the intro section of the ndarray doc page. We can point out that arrays are made up of scalars, and 0-D array (scalar array) is an ndarray instance containing precisely one array scalar with @EricCousineau-TRI's example added to the Example section. Such an explanation also provides some context for users going into the Scalars page as they see the connection between arrays and scalars.

**Comment by seberg** [comment]
To be honest, because i had the confusion again today (I am not sure if me or not!)... And scalar-arrays vs. array-scalar is just annoying (because I at least sometimes use scalar-array for 0-D array, because the behave like a scalar)... Maybe we should just call it "NumPy scalar".

**Comment by charris** [comment]
Given that I had to search the docs to see what the term meant, and it wasn't what I thought is was, we should probably use a clearer term.

**Comment by mattip** [comment]
> 0-dimensional array is synonymous with scalar array. 

"Scalar array" only appears in the documentation in old release notes and in the swig interface documentation, so I think we should not encourage the use of that imprecise term. We have "array scalars" (which IMO should be called "numpy scalars" since they are not containers) and "0d arrays".

**Comment by rgommers** [comment]
gh-18965 improves the documentation in the glossary of "array scalar".

@jasmincl, @TheCheshireCat6 and @mattip, you discussed some other changes related to this issue, right? Could one of you summarize those?

**Comment by mattip** [comment]
As long as we have 0-d arrays, we should have a clear explaination of what they are along the lines of 
> It appears that 0-D array is briefly explained in https://numpy.org/devdocs/reference/arrays.ndarray.html#calculation. However, I think that should be mentioned earlier in the intro section of the ndarray doc page.

**Comment by miccoli** [comment]
> Array-scalars are the scalar versions, which unlike arrays are immutable. But it is important to note that NumPy is very agressive at turning scalar arrays into array scalars. (i.e. turn 0-D arrays into scalars.) That behaviour can be bothersome if you are working with `object` arrays.

I learnd this behaviour the hard way, i.e. by trial and error. E.g.
```
>>> a = np.array(1)
>>> a *= 10
>>> a
array(10)
>>> a[()] = a * 10
>>> a
array(100)
>>> a = a * 10
>>> a
1000
```

Athough I can understand the reasons for which `a * 10` is a scalar and not a 0-D array, the surprise factor is high, and should be documented.

On the contrary I never fully understood the reason for which ufuncs with a 0-D array argument return a scalar and not a 0-D array:
```
>>> np.cos(np.array(0))
1.0
```
but again, once you understand how it works, I never found an inconsistency.

Summing up: current behaviour is (as far as I understand) consistent but poorly documented.

Cross reference: #4563 and #6248
